filewatcher/v2/filewatcher.go

```diff
@@ -0,0 +1,567 @@
+// it blocks until the file becomes available, or context is cancelled,
+// whichever comes first.
+//
+// When logger is non-nil, it will be used to log errors,
+// either returned by parser or by the underlying file system watcher.
+// Please note that this does not include errors returned by the first parser
+// call, which will be returned directly.
+func New[T any](ctx context.Context, path string, parser Parser[T], options ...Option) (*Result[T], error) {
+	opt := opts{
+		fsEventsDelay:       DefaultFSEventsDelay,
+		pollingInterval:     DefaultPollingInterval,
+		initialReadInterval: DefaultInitialReadInterval,
+
+		fileSizeLimit: DefaultMaxFileSize,
+	}
+	for _, o := range options {
+		o(&opt)
+	}
+	hardLimit := opt.fileSizeLimit * HardLimitMultiplier
+
+	var data T
+	var mtime time.Time
+	var files []string
+
+	var lastErr error
+	for {
+		select {
+		default:
+		case <-ctx.Done():
+			return nil, fmt.Errorf(
+				"filewatcher: context canceled while waiting for file(s) under %q to load: %w, last err: %w",
+				path,
+				ctx.Err(),
+				lastErr,
+			)
+		}
+
+		var err error
+		data, mtime, files, err = openAndParse(path, parser, opt.fileSizeLimit, hardLimit)
+		if errors.Is(err, fs.ErrNotExist) {
+			lastErr = err
+			time.Sleep(opt.initialReadInterval)
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		break
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	for _, path := range files {
+		if err := watcher.Add(path); err != nil {
+			return nil, fmt.Errorf(
+				"filewatcher: failed to watch %q: %w",
+				path,
+				err,
+			)
+		}
+	}
+
+	res := new(Result[T])
+	res.data.Store(&atomicData[T]{
+		data:  data,
+		mtime: mtime,
+	})
+	res.ctx, res.cancel = context.WithCancel(context.Background())
+
+	go res.watcherLoop(
+		watcher,
+		path,
+		parser,
+		opt.fileSizeLimit,
+		hardLimit,
+		opt.pollingInterval,
+		opt.fsEventsDelay,
+	)
+
+	return res, nil
+}
+
+// NewMockFilewatcher returns a pointer to a new MockFileWatcher object
+// initialized with the given io.Reader and Parser.
+func NewMockFilewatcher[T any](r io.Reader, parser Parser[T]) (*MockFileWatcher[T], error) {
+	fw := &MockFileWatcher[T]{parser: parser}
+	if err := fw.Update(r); err != nil {
+		return nil, err
+	}
+	return fw, nil
+}
+
+// MockFileWatcher is an implementation of FileWatcher that does not actually read
+// from a file, it simply returns the data given to it when it was initialized
+// with NewMockFilewatcher. It provides an additional Update method that allows
+// you to update this data after it has been created.
+type MockFileWatcher[T any] struct {
+	data   atomic.Value
+	parser Parser[T]
+}
+
+// Update updates the data of the MockFileWatcher using the given io.Reader and
+// the Parser used to initialize the FileWatcher.
+//
+// This method is not threadsafe.
+func (fw *MockFileWatcher[T]) Update(r io.Reader) error {
+	data, err := fw.parser(r)
```

Comment:
this is a [fake, not a mock](https://martinfowler.com/bliki/TestDouble.html)

Reply:
This took me some time but now I'm fully sold on mock vs fakes.